File: apps/daemon/src/identity.ts
/**
 * Device identity — stable across reinstalls.
 *
 * Lives at ~/.modelstat/identity.json. Contains the fields a fresh
 * install needs to resume a previous enrollment without triggering
 * a new self-register (which would create a ghost device row in the
 * user's account).
 *
 * Lives under the single daemon home (`modelstatHome()`, default `~/.modelstat`,
 * relocatable via MODELSTAT_HOME) alongside `state.json` — see ./paths.ts. The
 * identity file holds a long-lived bearer, so it gets explicit chmod 0600.
 * Runtime bookkeeping (cursors, API-URL override) lives in the sibling
 * `state.json` (./runtime-state.ts); both share one location so state can't fork.
 */

import {
  chmodSync,
  mkdirSync,
  readFileSync,
  renameSync,
  writeFileSync,
  existsSync,
} from "node:fs";
import { hostname as osHostname } from "node:os";
import { join } from "node:path";
import { modelstatHome } from "./paths.js";

export interface DeviceIdentity {
  /** Daemon-generated UUIDv7. Stable across reinstalls. */
  deviceUuid: string;
  /** Server UUID returned from /v1/devices/self-register. */
  deviceId: string;
  /** Long-lived Bearer (ds_live_…). Rotate via /v1/devices/me/rotate-secret. */
  bearerToken: string;
  /** Claim code surfaced to the user. Once the device is claimed, this stays
   * non-null but the server returns `status: "claimed"` on /devices/me. */
  claimCode: string | null;
  claimUrl: string | null;
  /** Hostname at the moment of self-register. Shown in the "reuse
   * existing device?" prompt so the user can spot if they're about to
   * reuse an identity that belongs to a different machine. */
  hostname: string;
  /** ISO timestamp. For audit + UI. */
  createdAt: string;
  /** Optional — set once the daemon discovers from /devices/me that the
   * device is claimed. Just a display convenience. */
  userEmail?: string | null;
  /** Default routing org snapshot. Display-only. */
  defaultOrgId?: string | null;
}

const IDENTITY_FILE = join(modelstatHome(), "identity.json");

function ensureRoot(): void {
  mkdirSync(modelstatHome(), { recursive: true, mode: 0o700 });
}

/** Atomic write + chmod 0600. Refuses to overwrite silently — caller
 * should call `backupIdentity()` first if an existing file is present
 * and represents a different identity. */
function writeAtomic(meta: DeviceIdentity): void {
  ensureRoot();
  const tmp = `${IDENTITY_FILE}.${process.pid}.tmp`;
  writeFileSync(tmp, JSON.stringify(meta, null, 2), { mode: 0o600 });
  renameSync(tmp, IDENTITY_FILE);
  // rename preserves tmp's mode; extra chmod ensures it in case
  // umask altered it on some platforms.
  try {
    chmodSync(IDENTITY_FILE, 0o600);
  } catch {
    /* best effort */
  }
}

export function identityPath(): string {
  return IDENTITY_FILE;
}

export function hasIdentityFile(): boolean {
  return existsSync(IDENTITY_FILE);
}

function parseFile(): DeviceIdentity | null {
  try {
    const raw = readFileSync(IDENTITY_FILE, "utf8");
    const obj = JSON.parse(raw) as Partial<DeviceIdentity>;
    if (!obj.deviceUuid || !obj.deviceId || !obj.bearerToken) {
      return null;
    }
    return {
      deviceUuid: obj.deviceUuid,
      deviceId: obj.deviceId,
      bearerToken: obj.bearerToken,
      claimCode: obj.claimCode ?? null,
      claimUrl: obj.claimUrl ?? null,
      hostname: obj.hostname ?? osHostname(),
      createdAt: obj.createdAt ?? new Date().toISOString(),
      userEmail: obj.userEmail ?? null,
      defaultOrgId: obj.defaultOrgId ?? null,
    };
  } catch {
    return null;
  }
}

/** Read the canonical identity from `identity.json`, or null if absent. */
export function loadIdentity(): DeviceIdentity | null {
  return parseFile();
}

/** Persist the current identity. Writes atomically; caller is
 * responsible for deciding whether an existing file should be backed
 * up first (e.g., user chose `--fresh`). */
export function saveIdentity(meta: DeviceIdentity): void {
  writeAtomic(meta);
}

/**
 * Rename the current identity file to
 * `~/.modelstat/identity.json.bak-<ISO timestamp>`. Returns the backup
 * path if one was made, or null if no file existed to back up.
 */
export function backupIdentity(): string | null {
  if (!existsSync(IDENTITY_FILE)) return null;
  const stamp = new Date().toISOString().replace(/[:.]/g, "-");
  const dest = `${IDENTITY_FILE}.bak-${stamp}`;
  renameSync(IDENTITY_FILE, dest);
  return dest;
}

/** Partial update — read + merge + write. Use for stamping
 * userEmail / defaultOrgId / claim state changes without touching
 * the device_uuid + bearer. */
export function updateIdentity(patch: Partial<DeviceIdentity>): DeviceIdentity | null {
  const current = parseFile();
  if (!current) return null;
  const merged: DeviceIdentity = { ...current, ...patch };
  writeAtomic(merged);
  return merged;
}

File: daemon/crates/modelstat-ingest/src/identity.rs
//! Device identity store — `~/.modelstat/identity.json`. Byte-for-byte port of
//! the TS `apps/daemon/src/identity.ts` (feature §4). Holds the long-lived bearer
//! (`ds_live_…`), so it is written atomically with `0600` perms; a fresh install
//! reuses it to resume an enrollment instead of minting a duplicate device row.

use std::path::PathBuf;

use serde::{Deserialize, Serialize};

use crate::paths::{ensure_home, home_path};
use crate::timefmt::now_iso;

/// The canonical identity record. Serializes to the exact camelCase keys, in the
/// exact order, of the TS `DeviceIdentity` (the `identity.json` contract + the
/// standalone MCP package which reads this file). `claimCode`/`claimUrl`/
/// `userEmail`/`defaultOrgId` are nullable — present as `null` when unset.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeviceIdentity {
    pub device_uuid: String,
    pub device_id: String,
    pub bearer_token: String,
    pub claim_code: Option<String>,
    pub claim_url: Option<String>,
    pub hostname: String,
    pub created_at: String,
    pub user_email: Option<String>,
    pub default_org_id: Option<String>,
}

/// Tolerant parse target — every field optional so a hand-edited or partial file
/// deserializes; validation (the three required fields) happens in `parse_file`.
#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
struct PartialIdentity {
    device_uuid: Option<String>,
    device_id: Option<String>,
    bearer_token: Option<String>,
    claim_code: Option<String>,
    claim_url: Option<String>,
    hostname: Option<String>,
    created_at: Option<String>,
    user_email: Option<String>,
    default_org_id: Option<String>,
}

/// `~/.modelstat/identity.json` (honors `MODELSTAT_HOME`; computed per call).
pub fn identity_path() -> PathBuf {
    home_path("identity.json")
}

/// Whether the identity file exists on disk.
pub fn has_identity_file() -> bool {
    identity_path().exists()
}

use crate::machine_key::host_name;

/// Parse the file. Requires `deviceUuid` + `deviceId` + `bearerToken`; anything
/// else defaults (hostname → this host, createdAt → now, the rest → null). A
/// missing/corrupt/incomplete file reads as `None` — treated as "no identity",
/// exactly like the TS `parseFile()`.
fn parse_file() -> Option<DeviceIdentity> {
    let raw = std::fs::read_to_string(identity_path()).ok()?;
    let obj: PartialIdentity = serde_json::from_str(&raw).ok()?;
    let (device_uuid, device_id, bearer_token) =
        match (obj.device_uuid, obj.device_id, obj.bearer_token) {
            (Some(u), Some(i), Some(b)) if !u.is_empty() && !i.is_empty() && !b.is_empty() => {
                (u, i, b)
            }
            _ => return None,
        };
    Some(DeviceIdentity {
        device_uuid,
        device_id,
        bearer_token,
        claim_code: obj.claim_code,
        claim_url: obj.claim_url,
        hostname: obj.hostname.unwrap_or_else(host_name),
        created_at: obj.created_at.unwrap_or_else(now_iso),
        user_email: obj.user_email,
        default_org_id: obj.default_org_id,
    })
}

/// Read the canonical identity, or `None` if absent/invalid.
pub fn load_identity() -> Option<DeviceIdentity> {
    parse_file()
}

/// Atomic write + `0600` (Unix). Writes `<file>.<pid>.tmp`, then renames — the
/// file is never observed half-populated. Caller decides whether to
/// `backup_identity()` first. Port of the TS `writeAtomic`.
pub fn save_identity(meta: &DeviceIdentity) -> std::io::Result<()> {
    ensure_home()?;
    let path = identity_path();
    let tmp = crate::atomic::with_pid_tmp(&path);
    let json = serde_json::to_string_pretty(meta).expect("DeviceIdentity always serializes");
    std::fs::write(&tmp, json)?;
    crate::atomic::set_file_0600(&tmp);
    std::fs::rename(&tmp, &path)?;
    // rename preserves the tmp mode; re-assert in case umask altered it.
    crate::atomic::set_file_0600(&path);
    Ok(())
}

/// Rename the current identity file to `identity.json.bak-<ISO ts, ':'/'.'→'-'>`.
/// Returns the backup path, or `None` if there was nothing to back up. Port of
/// the TS `backupIdentity`.
pub fn backup_identity() -> Option<PathBuf> {
    let path = identity_path();
    if !path.exists() {
        return None;
    }
    let stamp = now_iso().replace([':', '.'], "-");
    let dest = crate::atomic::sibling(&path, &format!("identity.json.bak-{stamp}"));
    std::fs::rename(&path, &dest).ok()?;
    Some(dest)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_env_lock;

    fn with_home<T>(dir: &std::path::Path, f: impl FnOnce() -> T) -> T {
        let _g = test_env_lock();
        std::env::set_var("MODELSTAT_HOME", dir);
        let out = f();
        std::env::remove_var("MODELSTAT_HOME");
        out
    }

    fn sample() -> DeviceIdentity {
        DeviceIdentity {
            device_uuid: "6f9619ff-8b86-5d11-b42d-00cf4fc964ff".into(),
            device_id: "device-server-uuid-abc".into(),
            bearer_token: "ds_live_examplefake0000000000000000000000000000".into(),
            claim_code: Some("brave-otter-lake".into()),
            claim_url: Some("https://modelstat.ai/claim/brave-otter-lake".into()),
            hostname: "marks-mac.local".into(),
            created_at: "2026-06-01T10:00:00.000Z".into(),
            user_email: None,
            default_org_id: None,
        }
    }

    #[test]
    fn save_then_load_roundtrips() {
        let tmp = tempfile::tempdir().unwrap();
        with_home(tmp.path(), || {
            save_identity(&sample()).unwrap();
            assert!(has_identity_file());
            assert_eq!(load_identity().unwrap(), sample());
        });
    }

    #[test]
    fn serialization_matches_golden_shape() {
        // Semantic equality against the committed file-format golden (order- and
        // whitespace-independent) + a key-order spot check.
        let golden_path = format!(
            "{}/../modelstat-wire/tests/golden/file_formats/identity.json",
            env!("CARGO_MANIFEST_DIR")
        );
        let golden: serde_json::Value =
            serde_json::from_str(&std::fs::read_to_string(&golden_path).unwrap()).unwrap();
        let mine: serde_json::Value = serde_json::to_value(sample()).unwrap();
        assert_eq!(
            mine, golden,
            "identity.json shape drifted from the TS golden"
        );

        let s = serde_json::to_string_pretty(&sample()).unwrap();
        let du = s.find("deviceUuid").unwrap();
        let di = s.find("deviceId").unwrap();
        let bt = s.find("bearerToken").unwrap();
        assert!(
            du < di && di < bt,
            "key order must match the TS object literal"
        );
    }

    #[test]
    fn missing_required_field_reads_as_none() {
        let tmp = tempfile::tempdir().unwrap();
        with_home(tmp.path(), || {
            crate::paths::ensure_home().unwrap();
            std::fs::write(identity_path(), r#"{"deviceUuid":"u","deviceId":"d"}"#).unwrap();
            assert!(
                load_identity().is_none(),
                "no bearerToken ⇒ treated as absent"
            );
        });
    }

    #[test]
    fn backup_renames_with_stamp() {
        let tmp = tempfile::tempdir().unwrap();
        with_home(tmp.path(), || {
            save_identity(&sample()).unwrap();
            let bak = backup_identity().expect("a file existed");
            assert!(!has_identity_file(), "original moved aside");
            let name = bak.file_name().unwrap().to_string_lossy();
            assert!(name.starts_with("identity.json.bak-"));
            // The timestamp segment after `bak-` has had every ':' and '.'
            // replaced with '-' (feature §4 backup naming).
            let stamp = name.strip_prefix("identity.json.bak-").unwrap();
            assert!(!stamp.contains(':') && !stamp.contains('.'));
        });
    }
}
